Setting up Hydra on Google Cloud instances

tahir24434 opened the pull request:
## Script pre-reqs:

sudo apt-get install python-pip
sudo pip install shell_command google-api-python-client fabric
## To setup, run

python mesos_marathon_setup.py
python hydra_setup_script.py
## TODOs:

-- Remove aurora dependency (Functions to remove this dependency are already there. They need testing though)
-- Upload slave configuration with correct TAG.
-- Write master script.

Commit:
WIP: Mesos master and marathon setup on gcloud

TODO: Slave configuration still needs to be done.
      Few Aurora commands are used. In opensource, people does not have access to Aurora.
      So, these aurora commands need to be replaced.
      Error handling needs to be done.

Signed-off-by: trauf <tahir24434@gmail.com>

tahir24434 committed:
Making script more resilient and readable

Introduced Fabric library to run ssh commands.
Wrote wrappers on top of fabric library.

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Hydra configuration

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
small fixes

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Installtion of libzmq

Slaves required libzmq to run tests.

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Minor fixes

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Some cleanup

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Minor fixes

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
minor fix

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Splitting up setup script into multiple scrtips. Fixes.

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

kratos7 commented on src/main/scripts/setup/test.py:
```diff
@@ -0,0 +1,20 @@
+instances = list_instances(compute, "festive-courier-755", "us-central1-f", filt)
+print "Name             Status            IP"
+for instance in instances:
+  if instance["name"].startswith('tahir'):
```
:)

sushilks commented on src/main/scripts/setup/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,215 @@
+
+parser = argparse.ArgumentParser(description='Mesos Marathon setup script')
+# 'default=3' fulfills the Apache Mesos recommendation of having at least three masters for a production environment.
+parser.add_argument('--config_file', '-f', type=str, default=os.getcwd() + "/setup_config.ini", help='Setup configuration file')
```
Add support for tags so multiple setups can be used.

tahir24434 replied:
Sure. I will add that.

sushilks commented on src/main/scripts/setup/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,215 @@
+# 'default=3' fulfills the Apache Mesos recommendation of having at least three masters for a production environment.
+parser.add_argument('--config_file', '-f', type=str, default=os.getcwd() + "/setup_config.ini", help='Setup configuration file')
+
+parser.add_argument('--local_work_dir', '-l', type=str, default=os.environ['HOME'], help='Script will copy all downloaded/output files in this directory')
```
If not specified you can either choose current directory or put it in a hidden directory in home 
i.e. $HOME/.hydra

sushilks commented on src/main/scripts/setup/setup_helpers.py:
```diff
@@ -0,0 +1,205 @@
+from tempfile import mkstemp
+from fabric.api import *
+
+project="festive-courier-755"
```
these values should come from the ini file, in most cases you dont need to specify as the default install of google (i.e. goggle init) already has them cached.

tahir24434 committed:
Minor fix

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Added correct grouping information. Bug fixes

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Removing un-necessary file

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

sushilks commented:
Add the following to build.py 
+    project.build_depends_on('shell_command')
+    project.build_depends_on('google-api-python-client')
+    project.build_depends_on('fabric')

sushilks commented on src/main/scripts/setup/hydra_setup_script.py:
```diff
@@ -0,0 +1,75 @@
+import os
```
Can we have only one launcher script ? (Similar to aurora)
you can call other functions internally from this script and the help on this script can describe all the other scripts.

tahir24434 replied:
Working on it as high priority.

sushilks commented on src/main/scripts/setup/hydra_setup_script.py:
```diff
@@ -0,0 +1,75 @@
+parser = argparse.ArgumentParser(description='Hydra setup script')
+parser.add_argument('--setup_ips_dir', '-i', type=str, default=os.environ['HOME'], help='Directory where mesos_masters_ips, mesos_slaves_ips and mesos_all_ips files are located. Default is /home/$USER')
+
+parser.add_argument('--dst_user_name', '-u', type=str, default="plumgrid", help='Destination user name')
```
default="plumgrid" is not good, use environment variable to get current user.

tahir24434 replied:
dst_user_name is user at cloud instance.

sushilks replied:
yes but you should remove plumgrid

tahir24434 replied:
Done.

sushilks commented on src/main/scripts/setup/hydra_setup_script.py:
```diff
@@ -0,0 +1,75 @@
+parser.add_argument('--setup_ips_dir', '-i', type=str, default=os.environ['HOME'], help='Directory where mesos_masters_ips, mesos_slaves_ips and mesos_all_ips files are located. Default is /home/$USER')
+
+parser.add_argument('--dst_user_name', '-u', type=str, default="plumgrid", help='Destination user name')
+parser.add_argument('--dst_work_dir', '-w', type=str, default="/home/plumgrid", help='Destination work directory. All contents will be uploaded here.')
```
Again default='/home/plumgrid' is not good, use the username extracted from environment.

tahir24434 replied:
dst_work_dir is directory at cloud instance. If something needs to be downloaded, it would be downloaded in that directory.

tahir24434 replied:
Done.

sushilks commented on src/main/scripts/setup/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,212 @@
+parser.add_argument('--config_file', '-f', type=str, default=os.getcwd() + "/setup_config.ini", help='Setup configuration file')
+
+parser.add_argument('--local_work_dir', '-l', type=str, default=os.environ['HOME'], help='Script will copy all downloaded/output files in this directory')
+parser.add_argument('--dst_work_dir', '-w', type=str, default="/home/plumgrid", help='Destination work directory. All contents will be uploaded here.')
```
remove plumgrid

tahir24434 replied:
Done

sushilks commented on src/main/scripts/setup/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,212 @@
+
+parser.add_argument('--local_work_dir', '-l', type=str, default=os.environ['HOME'], help='Script will copy all downloaded/output files in this directory')
+parser.add_argument('--dst_work_dir', '-w', type=str, default="/home/plumgrid", help='Destination work directory. All contents will be uploaded here.')
+parser.add_argument('--dst_user_name', '-u', type=str, default="plumgrid", help='Destination user name')
```
remove plugmrid.

tahir24434 replied:
Done.

sushilks commented on src/main/scripts/setup/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,212 @@
+      machinetype = options_dict["machinetype"]
+      tag = options_dict["tag"]
+      for i in range(int(count)):
+        cmd = "aurora spawn " + section + "-" + tag + "-" + str(i) + " ubuntu-14-04 " + machinetype
```
can't use aurora here.

tahir24434 replied:
Done

sushilks commented on src/main/scripts/setup/setup_helpers.py:
```diff
@@ -0,0 +1,228 @@
+    ips.append(instance["networkInterfaces"][0]["networkIP"])
+  return ips
+
+def spawn_instance(instance_name, os_name, machine_type="n1-standard-4", dst_user="plumgrid"):
```
remove defaults.

tahir24434 replied:
Done.

sushilks commented on src/main/scripts/setup/setup_helpers.py:
```diff
@@ -0,0 +1,228 @@
+  tfile = open(pathname, 'w')
+  #tfile = os.fdopen(fd, "w")
+
+  with open("/home/muneeb/.ssh/id_rsa.pub") as f:
```
muneeb ??? he is famous :-)
All key information should come from config file.

tahir24434 replied:
lol ... I will clean this up. This function is not currently being used. We will remove aurora dependency by using this function.

tahir24434 replied:
Done.

sushilks commented on src/main/scripts/setup/setup_helpers.py:
```diff
@@ -0,0 +1,228 @@
+  #shell_call(key_cmd)
+
+  print ("Creating the disk[%s-d1] for the instance" % instance_name)
+  disk1_cmd="gcloud compute disks create " + instance_name + "-d1 --image " + os_name + " --type pd-standard --size=30GB -q"
```
disk size also should come from config file.

sushilks commented on src/main/scripts/setup/setup_helpers.py:
```diff
@@ -0,0 +1,228 @@
+  print("disk1_cmd=%s" %disk1_cmd)
+  shell_call(disk1_cmd)
+  print ("Creating the disk[%s-d2] for the instance" % instance_name)
+  disk2_cmd="gcloud compute disks create " + instance_name + "-d2 --type pd-standard --size=75GB -q"
```
Same for this disk size. Get it from config file.

sushilks commented on src/main/scripts/setup/setup_helpers.py:
```diff
@@ -0,0 +1,228 @@
+  shell_call(disk2_cmd)
+  print ("Creating the instance[instance_name] ")
+  cmd = "gcloud compute instances create " + instance_name + " --machine-type " + machine_type + \
+        " --network net-10-10 --maintenance-policy MIGRATE --scopes https://www.googleapis.com/auth/cloud-platform --disk name=" + \
```
Network name should also come from config file.

sushilks commented:
Great work, it's coming together nicely. 
Just needs some minor cleanup.

Some tools were broken on the hydra side so it did not lint the commit but it should work in your next commit.

sushilks commented:
There are about 245 lint error's in the new code, however the command "pyb analyze" does not work on the directory inside src/main/scripts/setup/
Consider moving the python source from src/main/scripts/setup  to src/main/python/gcloud

tahir24434 committed:
Addressing comments

- Remove aurora dependency
- Get value of project/zone/email from configuration file
- Support for deployment tags
- Bug fixes

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Removing last traces of aurora. Bug fixes

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Code passing lint errors

Code is now in right directory.
Bug fixex

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Installtion librmq pre-reqs. Addressing comments

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+                    default=os.getcwd() + "/setup_config.ini", help='Setup configuration file')
+parser.add_argument('--deployment_id', '-i', type=str,
+                    help='Each cluster deployment needs to have a unique identifier. '
+                         'This will help in identifying various deployments.', required=True)
```
It does display as "optional" when you launch the script

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+parser = argparse.ArgumentParser(description='Mesos Marathon setup script')
+parser.add_argument('--config_file', '-f', type=str,
+                    default=os.getcwd() + "/setup_config.ini", help='Setup configuration file')
+parser.add_argument('--deployment_id', '-i', type=str,
```
Write that you can't use underscore char

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+                    help='Each cluster deployment needs to have a unique identifier. '
+                         'This will help in identifying various deployments.', required=True)
+
+parser.add_argument('--local_work_dir', '-l', type=str, default=os.environ['HOME'],
```
Personal opinion, I would not expose this

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+
+parser.add_argument('--local_work_dir', '-l', type=str, default=os.environ['HOME'],
+                                        help='Script will copy all downloaded/output files in this directory')
+parser.add_argument('--ssh_key_file', '-k', type=str, default=os.environ['HOME'] + "/.ssh/id_rsa.pub",
```
I would display what's going to happen if you don't specify one (e.g. 'default value is ~/.ssh/id_rsa')

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+                                        help='Script will copy all downloaded/output files in this directory')
+parser.add_argument('--ssh_key_file', '-k', type=str, default=os.environ['HOME'] + "/.ssh/id_rsa.pub",
+                    help='SSH public key absolute path. It would be used to get passwordless login to cloud instances')
+parser.add_argument('--start', '-r', type=int, default=1, help='start step')
```
Personal opinion, I would not expose this

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+# Zookeeper to keep track of the current leader of the master servers.
```
I would suggest to integrate the setup script with the hydra command line.
Something like "hydra setup" should launch this script.

lorenzodavid commented on src/main/python/gcloud_hydra/mesos_marathon_setup.py:
```diff
@@ -0,0 +1,235 @@
+        #                    infra script along with mesos setup.
+        print "==> Writing mesos masters ips in " + local_work_dir + "/.mesos_masters_ips files"
+        master_ips = setup_helpers.get_master_instances_ips(deployment_id, config)
+        f = open(local_work_dir + '/.mesos_all_ips', 'w')
```
Instead of using "/", I would use "os.path.join"

tahir24434 committed:
Merge remote-tracking branch 'upstream/master'

tahir24434 committed:
fix.

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Fixed some lint errors

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>

tahir24434 committed:
Minor fix

Signed-off-by: tahir24434 <tahir.rauf1@gmail.com>
